src/main/scala/zio/interop/reactiveStreams/Adapters.scala

```diff
@@ -46,10 +40,157 @@ object Adapters {
     } yield (error, demandUnfoldSink(subscriber, demand))
 
   def publisherToStream[A](publisher: Publisher[A], bufferSize: Int): ZStream[Any, Throwable, A] =
-    Stream.unwrap(
-      QueueSubscriber.make[A](bufferSize).flatMap {
-        case (subscriber, stream) => UIO(publisher.subscribe(subscriber)).const(stream)
-      }
+    new ZStream(
+      for {
+        (q, subscription, completion, subscriber) <- makeSubscriber[A](bufferSize).toManaged_
+        _                                         <- UIO(publisher.subscribe(subscriber)).toManaged_
+        sub <- subscription.await.interruptible
+                .toManaged(sub => UIO { sub.cancel() }.whenM(completion.isDone.map(!_)))
+        process <- process(q, sub, completion)
+      } yield process
     )
 
+  def sinkToSubscriber[R, R1 <: R, A1, A, B](
+    sink: ZSink[R, Throwable, A1, A, B],
+    bufferSize: Int
+  ): ZIO[R1, Throwable, (Subscriber[A], IO[Throwable, B])] =
+    for {
+      (q, subscription, completion, subscriber) <- makeSubscriber[A](bufferSize)
+      result                                    <- Promise.make[Throwable, B]
+      fiber <- new ZStream(for {
+                sub <- subscription.await.interruptible
+                        .toManaged(sub => UIO { sub.cancel() }.whenM(completion.isDone.map(!_)))
+                process <- process(q, sub, completion)
+              } yield process)
+                .run(sink)
+                .to(result)
+                .onInterrupt(
+                  subscription.poll.flatMap {
+                    case None     => UIO.unit
+                    case Some(io) => io.map(_.cancel()).orElse(UIO.unit)
+                  } *> q.shutdown
+                )
+                .fork
+    } yield (subscriber, result.await.interruptible.onInterrupt(fiber.interrupt))
```

Comment:
`result.await.interruptible` is superfluous; it is interruptible by default.

I think making the outer effect a `ZManaged` that interrupts the fiber on exit would make this `onInterrupt` redundant.

Reply:
Right, looks much better that way.